File: Programmers/LEVEL_2/c++/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue<int> q;
    int now_weight = 0;

    for (int i = 0; i < truck_weights.size(); i++)
    {
        int temp = truck_weights[i];
        while (true)
        {
            if (q.empty())
            {
                q.push(temp);
                answer++;
                now_weight += temp;
                break;
            }
            else if (q.size() == bridge_length)
            {
                now_weight -= q.front();
                q.pop();
            }
            else
            {
                if (now_weight + temp > weight)
                {
                    q.push(0);
                    answer++;
                }
                else
                {
                    q.push(temp);
                    answer++;
                    now_weight += temp;
                    break;
                }
            }
        }
    }
    answer += bridge_length;
    return answer;
}

File: Programmers/LEVEL_2/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
  var answer = 0;

  let queue = [];
  let total_weight = 0;

  for (let i = 0; i < bridge_length; i++) {
    queue.push(0);
  }

  let now = truck_weights.shift();
  queue.push(now);
  queue.shift();
  total_weight += now;
  answer++;

  while (total_weight) {
    total_weight -= queue.shift();
    now = truck_weights.shift();
    if (now + total_weight <= weight) {
      queue.push(now);
      total_weight += now;
    } else {
      queue.push(0);
      truck_weights.unshift(now); // 다시 꺼낸 값을 넣어주는게 포인트였다.
    }
    answer++;
  }

  return answer;
}
